core/trino-main/src/main/java/io/trino/operator/output/BlockTypeDispatchingPositionsAppender.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Dispatches the {@link #append} and {@link #appendRle} methods to the {@link #delegate} depending on the input {@link Block} class.
+ * The {@link Block} is flattened if necessary so that the {@link #delegate} {@link PositionsAppender#append(IntArrayList, Block)}
+ * always gets flat {@link Block} and {@link PositionsAppender#appendRle(RunLengthEncodedBlock)} always gets {@link RunLengthEncodedBlock}
+ * with {@link RunLengthEncodedBlock#getValue()} being flat {@link Block}.
+ */
+public class BlockTypeDispatchingPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BlockTypeDispatchingPositionsAppender.class).instanceSize();
+
+    private final PositionsAppender delegate;
+
+    public BlockTypeDispatchingPositionsAppender(PositionsAppender delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        if (positions.isEmpty()) {
+            return;
+        }
+
+        if (source instanceof RunLengthEncodedBlock) {
+            delegate.appendRle(flatten((RunLengthEncodedBlock) source, positions.size()));
+        }
+        else if (source instanceof DictionaryBlock) {
+            appendDictionary(positions, (DictionaryBlock) source);
+        }
+        else {
+            delegate.append(positions, source);
+        }
+    }
+
+    private void appendDictionary(IntArrayList positions, DictionaryBlock source)
+    {
+        Block dictionary = source.getDictionary();
+
+        while (dictionary instanceof RunLengthEncodedBlock || dictionary instanceof DictionaryBlock) {
+            if (dictionary instanceof RunLengthEncodedBlock) {
+                // since at some level, dictionary contains only a single value, it can be flattened to rle
+                appendRle(flatten((RunLengthEncodedBlock) dictionary, positions.size()));
```

Comment:
since you call `appendRle`, you don't need to `flatten` anymore. `appendRle` does it already